File: frontend/js/share.js
// Copyright (C) 2021 Charles All rights reserved.
// Author: @Charles-1414
// License: GNU General Public License v3.0

var shareList = [];
var page = 1;
var pageLimit = 10;
var orderBy = "name";
var order = "asc";
var search = "";

function UpdateShareList() {
    $("#refresh-btn").html('<i class="fa fa-sync fa-spin"></i>');
    $.ajax({
        url: "/api/share",
        method: 'POST',
        async: true,
        dataType: "json",
        data: {
            operation: "list",
            page: page,
            pageLimit: pageLimit,
            orderBy: orderBy,
            order: order,
            search: search,
            userId: localStorage.getItem("userId"),
            token: localStorage.getItem("token")
        },
        success: function (r) {
            shareList = r.data;
            total = r.total;
            totalShare = r.totalShare;
            $("tbody tr").remove();

            for (var i = 0; i < shareList.length; i++) {
                btns = '<button type="button" class="btn btn-warning btn-sm" onclick="Unshare(\'' + shareList[i].shareCode + '\')"><i class="fa fa-trash"></i></button>';
                var time = new Date(parseInt(shareList[i].timestamp) * 1000);
                AppendTableData("shareList", [shareList[i].name, shareList[i].shareCode, shareList[i].importCount, time.toLocaleString(), btns], shareList[i].bookId);
            }
            l = (page - 1) * pageLimit + 1;
            r = l + shareList.length - 1;
            if (shareList.length == 0) {
                AppendTableData("shareList", ["No data available"], undefined, "100%");
                l = 0;
            }

            SetTableInfo("shareList", "<p style='opacity:80%'>Showing " + l + " - " +
                r + " / " + totalShare);
            PaginateTable("shareList", page, total, "ShareListPage");

            $("#refresh-btn").html('<i class="fa fa-sync"></i>');
        },
        error: function (r, textStatus, errorThrown) {
            AjaxErrorHandler(r, textStatus, errorThrown);
        }
    });
}

function ShareListPage(p) {
    page = p;
    UpdateShareList();
}

function Search() {
    search = $("#search-content").val();
    orderBy = "none";
    SortTable(orderBy);
    UpdateShareList();
}

function UserListSort(id) {
    orderBy = id;
    if ($("#sorting_" + id).hasClass("sorting-desc")) {
        order = "desc";
    } else {
        order = "asc";
    }
    UpdateShareList();
}

function UpdatePageLimit(pl) {
    if (pageLimit != pl) {
        page = Math.ceil((page - 1) * pageLimit / pl + 1);
        pageLimit = pl;
        UpdateShareList();
    }
}

function PageInit() {
    UpdateShareList();
}

function Unshare(shareCode) {
    $.ajax({
        url: "/api/share",
        method: 'POST',
        async: true,
        dataType: "json",
        data: {
            operation: "remove",
            shareCode: shareCode,
            userId: localStorage.getItem("userId"),
            token: localStorage.getItem("token")
        },
        success: function (r) {
            if (r.success == true) {
                NotyNotification(r.msg, type = 'success');
                UpdateShareList();
            } else {
                NotyNotification(r.msg, type = 'error');
            }
        },
        error: function (r, textStatus, errorThrown) {
            AjaxErrorHandler(r, textStatus, errorThrown);
        }
    });
}

function CreateShareShow() {
    $("#content").after(`<div class="modal fade" id="modal" tabindex="-1" role="dialog"
        aria-labelledby="modalLabel" aria-hidden="true">
        <div class="modal-dialog" role="document">
            <div class="modal-content">
                <div class="modal-header">
                    <h5 class="modal-title"><i class="fa fa-share-alt"></i> Share Book
                    </h5>
                    <button type="button" class="close" style="background-color:transparent;border:none" data-dismiss="modal" aria-label="Close"
                        onclick="$('#modal').modal('hide')">
                        <span aria-hidden=" true"><i class="fa fa-times"></i></span>
                    </button>
                </div>
                <div class="modal-body">
                    <p>Select a book to share: </p>
                    <select name="bookId" id="book" class="form-select">
                    </select>
                    <p>An unique share code will be generated.</p>
                    <p><span id="share-msg"></span></p>
                </div>
                <div class="modal-footer">
                    <button type="button" class="btn btn-secondary" data-dismiss="modal"
                        onclick="$('#modal').modal('hide')">Cancel</button>
                    <button id="share-btn" type="button" class="btn btn-primary" data-dismiss="modal"
                        onclick="Share();">Share</button>
                </div>
            </div>
        </div>
    </div>`);
    $.ajax({
        url: "/api/book",
        method: 'POST',
        async: true,
        dataType: "json",
        data: {
            userId: localStorage.getItem("userId"),
            token: localStorage.getItem("token")
        },
        success: function (r) {
            for (var i = 1; i < r.length; i++)
                $("#book").append("<option value=" + r[i].bookId + ">" + r[i].name +
                    "</option>");
        }
    });
    $("#modal").modal("show");
    $('#modal').on('hidden.bs.modal', function () {
        $("#modal").remove();
    });
    setTimeout(function () {
        selected = [];
    }, 100);
}

function Share() {
    bookId = $('select[name="bookId"]').val();
    $.ajax({
        url: "/api/share",
        method: 'POST',
        async: true,
        dataType: "json",
        data: {
            operation: "create",
            bookId: bookId,
            userId: localStorage.getItem("userId"),
            token: localStorage.getItem("token")
        },
        success: function (r) {
            if (r.success == true) {
                NotyNotification(r.msg, type = 'success');
                $("#share-msg").html("Book share created! Share code: " + r.shareCode);
                UpdateShareList();
            } else {
                NotyNotification(r.msg, type = 'error');
            }
        },
        error: function (r, textStatus, errorThrown) {
            AjaxErrorHandler(r, textStatus, errorThrown);
        }
    });
}

File: backend/apis/share.py
# Copyright (C) 2021 Charles All rights reserved.
# Author: @Charles-1414
# License: GNU General Public License v3.0

from fastapi import Request, HTTPException
import os, sys, time
import json

from app import app, config
from db import newconn
from functions import *
import sessions

##########
# Share API

@app.post("/api/share")
async def apiShareBook(request: Request):
    ip = request.client.host
    form = await request.form()
    conn = newconn()
    cur = conn.cursor()
    if not "userId" in form.keys() or not "token" in form.keys() or "userId" in form.keys() and (not form["userId"].isdigit() or int(form["userId"]) < 0):
        raise HTTPException(status_code=401)

    userId = int(form["userId"])
    token = form["token"]
    if not validateToken(userId, token):
        raise HTTPException(status_code=401)
    
    op = form["operation"]

    if op == "list":
        page = int(form["page"])
        if page <= 0:
            page = 1

        pageLimit = int(form["pageLimit"])
        if pageLimit <= 0:
            pageLimit = 20
        
        if pageLimit > 100:
            return {"success": True, "data": [], "total": 0}

        orderBy = form["orderBy"]
        if not orderBy in ["name", "shareCode", "importCount", "timestamp"]:
            orderBy = "name"

        order = form["order"]
        if not order in ["asc", "desc"]:
            order = "asc"
        l = {"asc": 0, "desc": 1}
        order = l[order]

        search = form["search"]

        ret = []
        i2n = {}
        cur.execute(f"SELECT bookId, shareCode, importCount, createTS FROM BookShare WHERE userId = {userId} AND shareType = 0")
        d = cur.fetchall()
        for dd in d:
            name = ""
            if dd[0] in i2n.keys():
                name = i2n[dd[0]]
            else:
                cur.execute(f"SELECT name FROM Book WHERE bookId = {dd[0]}")
                t = cur.fetchall()
                if len(t) == 0:
                    continue
                name = decode(t[0][0])
                i2n[dd[0]] = name
            
            ret.append({"bookId": dd[0], "name": name, "shareCode": "!" + dd[1], "importCount": dd[2], "timestamp": dd[3]})
        
        t = {}
        i = 0
        for dd in ret:
            ok = False
            for tt in dd.keys():
                if search == "" or search in str(dd[tt]):
                    ok = True
                    break
            if not ok:
                continue
            i += 1
            t[str(dd[orderBy]) + str(dd["bookId"]) + str(i)] = dd
            
        ret = []
        for key in sorted(t.keys()):
            ret.append(t[key])
        if order == 1:
            ret = ret[::-1]

        if len(ret) <= (page - 1) * pageLimit:
            return {"success": True, "data": [], "total": (len(ret) - 1) // pageLimit + 1, "totalShare": len(ret)}
        elif len(ret) <= page * pageLimit:
            return {"success": True, "data": ret[(page - 1) * pageLimit :], "total": (len(ret) - 1) // pageLimit + 1, "totalShare": len(ret)}
        else:
            return {"success": True, "data": ret[(page - 1) * pageLimit : page * pageLimit], "total": (len(ret) - 1) // pageLimit + 1, "totalShare": len(ret)}

    elif op == "create":
        bookId = int(form["bookId"])

        cur.execute(f"SELECT * FROM Book WHERE userId = {userId} AND bookId = {bookId}")
        if len(cur.fetchall()) == 0:
            return {"success": False, "msg": "Book not found!"}
        
        cur.execute(f"SELECT shareCode FROM BookShare WHERE userId = {userId} AND bookId = {bookId} AND shareType = 0")
        t = cur.fetchall()
        if len(t) >= 20:
            return {"success": False, "msg": "You can create at most 20 shares for one book!"}
        else:
            shareCode = genCode(8)
            cur.execute(f"SELECT shareCode FROM BookShare WHERE shareCode = '{shareCode}'")
            if len(cur.fetchall()) != 0: # conflict
                for _ in range(30):
                    shareCode = genCode(8)
                    cur.execute(f"SELECT shareCode FROM Book WHERE shareCode = '{shareCode}'")
                    if len(cur.fetchall()) == 0:
                        break
                cur.execute(f"SELECT shareCode FROM Book WHERE shareCode = '{shareCode}'")
                if len(cur.fetchall()) != 0:
                    return {"success": False, "msg": "Unable to generate an unique share code..."}
                    
            cur.execute(f"INSERT INTO BookShare VALUES ({userId}, {bookId}, '{shareCode}', 0, {int(time.time())}, 0)")
            conn.commit()

            return {"success": True, "msg": "A new book share has been created!", "shareCode": "!" + shareCode}

    elif op == "remove":
        code = form["shareCode"]
        code = code.replace("!","").replace("@","")

        if not code.isalnum():
            return {"success": False, "msg": "Invalid share code!"}

        cur.execute(f"SELECT bookId FROM BookShare WHERE userId = {userId} AND shareType = 0 AND shareCode = '{code}'")
        t = cur.fetchall()
        if len(t) == 0:
            return {"success": False, "msg": "Invalid share code!"}
        else:
            cur.execute(f"DELETE FROM BookShare WHERE userId = {userId} AND bookId = {t[0][0]} AND shareCode = '{code}'")
            conn.commit()
            return {"success": True, "msg": "Book unshared!"}